pennylane/tape/tapes/reversible.py

```diff
@@ -180,26 +192,71 @@ def analytic_pd(self, idx, device, params=None, **options):
         else:
             generator, multiplier = op.generator
 
-        generator = generator(wires)
+        # construct circuit to compute differentiated state
+        between_ops_inverse = [copy.deepcopy(op) for op in between_ops[::-1]]
 
-        diff_circuit = QuantumTape()
-        diff_circuit._ops = [copy(op).inv() for op in between_ops[::-1]] + [generator] + between_ops
+        with QuantumTape() as new_circuit:
+            # start with final state of original circuit
+            qml.QubitStateVector(final_state, wires=dev_wires)
 
-        # set the simulator state to be the pre-measurement state
-        device._state = self._state
+            # evolve circuit backwards until gate we want to differentiate
+            for op in between_ops_inverse:
+                op.queue().inv()
 
-        # evolve the pre-measurement state under this new circuit
-        device.execute(diff_circuit)
-        dstate = device._pre_rotated_state  # TODO: this will only work for QubitDevices
+            # apply generator needed for differentiation
+            generator(wires)
 
-        # compute matrix element <d(state)|O|state> for each observable O
-        matrix_elems = device._asarray(
-            [self._matrix_elem(dstate, ob, self._state, device) for ob in self.observables]
-            # TODO: if all observables act on same number of wires, could
-            # do all at once with einsum
-        )
+            # evolve forwards again
+            for op in between_ops:
+                op.queue()
+
+            qml.tape.measure.state()
+
+        tapes = [new_circuit]
+
+        def processing_fn(results):
+            """Computes the gradient of the parameter at index idx via the
+            reversible backprop method.
+
+            Args:
+                results (list[real]): evaluated quantum tapes
+
+            Returns:
+                array[float]: 1-dimensional array of length determined by the tape output
+                measurement statistics
+            """
+            dstate = results[0][0]
+
+            # compute matrix element <d(state)|O|state> for each observable O
+            # TODO: if all observables act on same number of wires, could do all at once with einsum
+            matrix_elems = [
+                self._matrix_elem(dstate, ob, final_state, dev_wires) for ob in self.observables
+            ]
+            matrix_elems = np.array(matrix_elems)
+            return 2 * multiplier * np.imag(matrix_elems)
+
+        return tapes, processing_fn
+
+    def jacobian(self, device, params=None, **options):
+        # The parameter_shift_var method needs to evaluate the circuit

```

Comment:
Is `parameter_shift_var` the correct name for the function that needs the logic below?

Reply:
Fixed.